### PR TITLE
CCM-7251: Implement lambda authoriser

### PR DESCRIPTION
**Pull request by chris-elliott-nhsd**
## Description

PR to implement lambda authoriser for templates API

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: lambdas/authorizer/src/index.ts [hunk 1]
```diff
@@ -1,21 +1,95 @@
...
+      return generatePolicy(methodArn, 'Deny');
+    }
+
+    const decodedToken = jwtDecode(authorizationToken);
+
+    const {
+      client_id: clientId,
+      iss,
+      token_use: tokenUse,
+    } = $AccessToken.parse(decodedToken);
```
Comment (mark-r-bjss) [inline lambdas/authorizer/src/index.ts:57]: The token should have an expiry time and should be rejected if it is expired (although maybe that would happen before it gets here?)

File: lambdas/authorizer/src/index.ts [hunk 2]
```diff
@@ -1,21 +1,95 @@
...
+  try {
+    const userPoolId = getEnvironmentVariable('USER_POOL_ID');
+    const userPoolClientId = getEnvironmentVariable('USER_POOL_CLIENT_ID');
+
+    if (!userPoolId || !userPoolClientId) {
+      logger.error('Lambda misconfiguration');
+      return generatePolicy(methodArn, 'Deny');
+    }
+
+    const decodedToken = jwtDecode(authorizationToken);
```
Comment (mark-r-bjss) [inline lambdas/authorizer/src/index.ts:51]: Does this verify that signature (or has that already been done by AWS Gateway?)

File: lambdas/authorizer/src/index.ts [hunk 3]
```diff
@@ -1,21 +1,95 @@
...
+    // token_use claim
+    if (tokenUse !== 'access') {
+      logger.warn(
+        `Token has invalid token_use, expected access but received ${tokenUse}`
+      );
+      return generatePolicy(methodArn, 'Deny');
+    }
+
+    // cognito SDK call
+    await cognitoClient.send(
```
Comment (mark-r-bjss) [inline lambdas/authorizer/src/index.ts:84]: What is the purpose of this as we don't use the result?

File: lambdas/authorizer/package.json [hunk 4]
```diff
@@ -3,18 +3,28 @@
...
     "@types/jest": "^29.5.14",
+    "@types/jsonwebtoken": "^9.0.7",
     "aws-lambda": "^1.0.7",
     "jest": "^29.7.0",
     "jest-mock-extended": "^3.0.7",
     "typescript": "^5.5.4"
+  },
+  "dependencies": {
+    "@aws-sdk/client-cognito-identity-provider": "^3.693.0",
+    "esbuild": "^0.24.0",
```
Comment (harrim91) [inline lambdas/authorizer/package.json:24]: small thing - esbuild should be a dev dependency

Comment (chris-elliott-nhsd) [reply lambdas/authorizer/package.json]: updated

File: lambdas/authorizer/src/index.ts (hunk 2 above)
Comment (chris-elliott-nhsd) [reply lambdas/authorizer/src/index.ts]: The token was being verified in the SDK call however we have added an explicit verify step to the lambda as discussed.

File: lambdas/authorizer/src/index.ts (hunk 1 above)
Comment (chris-elliott-nhsd) [reply lambdas/authorizer/src/index.ts]: we are now verifying the token using `jwks-rsa`, which includes an expiry check

**Review by mark-r-bjss — APPROVED** [review]
Looks good, we need to make sure that we include some basic testing in the component tests once we have it deployed to a sandbox, it might be worth creating the associated test ticket as a reminder.

**Review by bhansell1 — APPROVED** [review]

**Review by mark-r-bjss — APPROVED** [review]

**Pull request merged**